physical/redis_test.go

```diff
@@ -0,0 +1,100 @@
+package physical
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"testing"
+	"time"
+	"github.com/garyburd/redigo/redis"
+)
+
+// To run this test, set up a local Redis instance and run:
+//
+//     docker run -p 6379:6379 -d redis
+//     env REDIS_URL=redis://localhost:6379/9 make test TEST=./physical
+func TestRedisBackend(t *testing.T) {
+	addr := os.Getenv("REDIS_URL")
+	if addr == "" {
+		t.SkipNow()
+	}
+
+	randPath := fmt.Sprintf("vault-leader-%d", time.Now().Unix())
+	// TODO: Delete all matching keys when done.
+
+	logger := log.New(os.Stderr, "", log.LstdFlags)
+
+	b, err := NewBackend("redis", logger, map[string]string{
+		"path": randPath,
+		"leader_ttl_renew_interval": "100",
+	})
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	testBackend(t, b)
+	testBackend_ListPrefix(t, b)
+
+	ha, ok := b.(HABackend)
+	if !ok {
+		t.Fatalf("redis does not implement HABackend")
+	}
+	testHABackend(t, ha, ha)
+
+	testRedisHAExpiration(t, ha)
+
+	// Clean up all unit test keys.
+	conn, err := redis.DialURL(addr)
+	defer conn.Close();
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+	reply, err := conn.Do("KEYS", randPath+"/*")
+	keys, err := redis.Strings(reply, err)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+	for _, key := range keys {
+		_, err = conn.Do("DEL", key)
+		if err != nil {
+			t.Fatalf("err: %s", err)
+		}
+	}
+}
+
+// Since RedisLock provides a SimulateExpiration method, we can test what
+// happens when the lock is lost unexpectedly.
+func testRedisHAExpiration(t *testing.T, b HABackend) {
+	// Get a lock.
+	lock, err := b.LockWith("foo", "bar")
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	redisLock, ok := lock.(*RedisLock)
+	if !ok {
+		t.Fatalf("redis did not return a RedisLock")
+	}
+
+	// Attempt to lock.
+	leaderCh, err := redisLock.Lock(nil)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	if leaderCh == nil {
+		t.Fatalf("failed to get leader ch")
+	}
+
```

Comment:
at this point in the test, you could simulate a second instance trying to get leadership and (properly) failing